File: src/main/java/com/github/hirsivaja/ip/ipv6/Ipv6Address.java
package com.github.hirsivaja.ip.ipv6;

import com.github.hirsivaja.ip.ByteArray;
import com.github.hirsivaja.ip.IpAddress;
import com.github.hirsivaja.ip.IpUtils;

import java.net.Inet6Address;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;

public record Ipv6Address(ByteArray address) implements IpAddress {
    public static final int IPV6_ADDRESS_LEN = 16;

    public Ipv6Address {
        if(address.array().length != IPV6_ADDRESS_LEN) {
            throw new IllegalArgumentException("Incorrect length for IPv6 address " + address.array().length);
        }
    }

    public Ipv6Address(byte[] addressBytes) {
        this(new ByteArray(addressBytes));
    }

    @Override
    public void encode(ByteBuffer out) {
        out.put(address.array());
    }

    @Override
    public int length() {
        return IPV6_ADDRESS_LEN;
    }

    @Override
    public byte[] rawAddress() {
        return address.array();
    }

    @Override
    public InetAddress toInetAddress() {
        try {
            return InetAddress.getByAddress(address.array());
        } catch (UnknownHostException e) {
            throw new IllegalArgumentException("Not a valid IPv6 address!", e);
        }
    }

    public Inet6Address toInet6Address() {
        return (Inet6Address) toInetAddress();
    }

    public static Ipv6Address decode(ByteBuffer in) {
        if(in.remaining() < IPV6_ADDRESS_LEN) {
            throw new IllegalArgumentException("Too few remaining bytes " + in.remaining());
        }
        byte[] addressBytes = new byte[IPV6_ADDRESS_LEN];
        in.get(addressBytes);
        return new Ipv6Address(addressBytes);
    }

    @Override
    public String toString() {
        try {
            return this.getClass().getSimpleName() + "[" + toInetAddress().getHostAddress() + "]";
        } catch (IllegalArgumentException _) {
            // Suppressing the exception
        }
        return this.getClass().getSimpleName() + "[" + IpUtils.printHexBinary(address.array()) + "]";
    }
}

File: src/main/kotlin/com/github/hirsivaja/ip/address/Ipv6Address.kt
package com.github.hirsivaja.ip.address

import com.github.hirsivaja.ip.DataArray
import java.net.InetAddress
import java.nio.ByteBuffer

data class Ipv6Address(val address: DataArray) : IpAddress {
    constructor(address: ByteArray) : this(DataArray(address))

    init {
        require(address.data.size == IPV6_ADDRESS_LEN) { "IPv6 address has to be 16 bytes long!" }
    }

    override fun encode(out: ByteBuffer) {
        out.put(address.data)
    }

    override fun length(): Int {
        return address.data.size
    }

    override fun toInetAddress(): InetAddress {
        return InetAddress.getByAddress(address.data)
    }

    override fun rawAddress(): ByteArray {
        return address.data
    }

    override fun toString(): String {
        val inetAddress = toInetAddress().toString().removePrefix("/")
        return "IPv6Address($inetAddress)"
    }

    companion object {
        fun decode(buf: ByteBuffer): Ipv6Address {
            val address = ByteArray(IPV6_ADDRESS_LEN)
            buf[address]
            return Ipv6Address(address)
        }

        const val IPV6_ADDRESS_LEN : Int = 16
    }
}
